mlflow/store/sftp_artifact_repo.py

```diff
@@ -0,0 +1,89 @@
+import os
+
+from mlflow.entities.file_info import FileInfo
+from mlflow.store.artifact_repo import ArtifactRepository
+from mlflow.utils.file_utils import TempDir
+
+try:
+    from urllib.parse import urlparse
+except ImportError:
+    from urlparse import urlparse
+
+
+class SFTPArtifactRepository(ArtifactRepository):
+    """Stores artifacts as files in a remote directory, via sftp."""
+
+    def __init__(self, artifact_uri, client=None):
+        self.uri = artifact_uri
+        parsed = urlparse(artifact_uri)
+        self.config = {
+            'host': parsed.hostname,
+            'port': 22 if parsed.port is None else parsed.port,
+            'username': parsed.username,
+            'password': parsed.password
+        }
+        self.path = parsed.path
+
+        if client:
+            self.sftp = client
+        else:
+            import pysftp
+            import paramiko
+
+            ssh_config = paramiko.SSHConfig()
+            user_config_file = os.path.expanduser("~/.ssh/config")
+            if os.path.exists(user_config_file):
+                with open(user_config_file) as f:
+                    ssh_config.parse(f)
+
+            user_config = ssh_config.lookup(self.config['host'])
+
+            if self.config['username'] is None and 'username' in user_config:
+                self.config['username'] = user_config['username']
+
+            if 'identityfile' in user_config:
+                self.config['private_key'] = user_config['identityfile'][0]
```

Comment:
Will `user_config['identityfile']` always be an array? Would be great to confirm if you're able to, the paramiko docs aren't super clear on this

Reply:
The ssh_config docs say:
> IdentityFile
> Specifies a file from which the user's RSA or DSA authentication identity is read. The default is ~/.ssh/identity for protocol version 1, and ~/.ssh/id_rsa and ~/.ssh/id_dsa for protocol version 2. Additionally, any identities represented by the authentication agent will be used for authentication.
>The file name may use the tilde syntax to refer to a user's home directory or one of the following escape characters: '%d' (local user's home directory), '%u' (local user name), '%l' (local host name), '%h' (remote host name) or '%r' (remote user name).
>It is possible to have multiple identity files specified in configuration files; all these identities will be tried in sequence.

So one would assume that IdentityFile will be an array. In my case, the returned array has length 1.